### PR TITLE
add type hints for probability distributions

### PR DESCRIPTION
**Pull request by joelberkeley-pio**
does #1368 for probability distributions

required a fix for `TensorLike` problem for multiple dispatch

### CODE REVIEW

**Comment by st--** [comment]
typing.Union[], unfortunately, does not play together with the multipledispatch. The multipledispatch package does take tuples; however, if this is used for several arguments this results in generating (number of explicit types in union-type) ^ (number of times this is used) elements. This may slow down dispatch. If we do decide to go for this I think we should include some dispatch timing tests to ensure this does not affect performance significantly.

**Comment by joelberkeley-pio** [comment]
> typing.Union[], unfortunately, does not play together with the multipledispatch, so as it stands this PR is not ready for merging. The multipledispatch package does take tuples; however, if this is used for several arguments this results in generating (number of explicit types in union-type) ^ (number of times this is used) elements. This may slow down dispatch. If we do decide to go for this I think we should include some dispatch timing tests to ensure this does not affect performance significantly.

true. multipledispatch doesn't work with `typing.Union`, but that's not why the build is failing. Line 8 in kufs.py
```
@Kuf.register(InducingPoints, Kernel, TensorLike)
```
worked before for `Parameter` because `TensorLike` was defined as `object`. Had it had been defined using the more specific `(tf.Tensor, tf.Variable, np.ndarray)` as suggested by the comments we'd see the above. In addition to your question about performance, there's also then the question: should `TensorLikeTypes` include `Parameter`, and if not what do we do with `Kuf`?

Regarding performance, would we expect multiple dispatch to be used in an optimization loop?

**Comment by st--** [comment]
> should `TensorLikeTypes` include `Parameter`?

Yes.

> Regarding performance, would we expect multiple dispatch to be used in an optimization loop?

Multiple dispatch is crucial for conditionals() & co which is at the core of every GPflow model... so yes. It may be fine, but I'd want to see a speed comparison before approving these changes.

**Comment by joelberkeley-pio** [comment]
How about I just do
```
TensorLikeTypes = (object,) # should be (tf.Tensor, tf.Variable, np.ndarray, Parameter) but waiting on performance test
```
for now? I'd rather not get bogged down in more details than is needed for now. We'd retain the same level of functionality as before, just with a few extra types

File: gpflow/base.py [hunk 1]
```diff
@@ -307,6 +304,9 @@ def _run_op(a, *args, **kwargs):
 Parameter._OverloadAllOperators()
 tf.register_tensor_conversion_function(Parameter, lambda x, *args, **kwds: x.read_value())
 
+TensorLike = Union[np.ndarray, tf.Tensor, tf.Variable, Parameter]
+TensorLikeTypes = (tf.Tensor, tf.Variable, np.ndarray, Parameter)
```
Comment (st--) [inline gpflow/base.py:308]: Would it not make more sense to have the types together at the top of the file (as it was before)?
Alternatively, it might be helpful to collate them in a `gpflow.types` module (but perhaps do that in a separate PR?)... :thinking:

Comment (st--) [reply gpflow/base.py:308]: It might also be helpful to add a comment explaining why we've got two things here that are almost exactly the same but not quite

**Comment by st--** [comment]
> How about I just do `TensorLikeTypes = (object,) # should be (tf.Tensor, tf.Variable, np.ndarray, Parameter) but waiting on performance test` for now?

I guess that could work. It's a bit unfortunate that it doesn't work as a type signature...!

File: gpflow/base.py (hunk 1 above)
Comment (joelberkeley-pio) [reply gpflow/base.py:308]: at the top of the file would be nice, but they need to be after the `Parameter` class to work

gpflow.types module? perhaps

comment? yes. will do

**Comment by joelberkeley-pio** [comment]
> > How about I just do `TensorLikeTypes = (object,) # should be (tf.Tensor, tf.Variable, np.ndarray, Parameter) but waiting on performance test` for now?
> 
> I guess that could work. It's a bit unfortunate that it doesn't work as a type signature...!

This is the risk that one has to accept by taking using third party utilities over the stdlib

File: gpflow/base.py [hunk 2]
```diff
@@ -307,6 +304,16 @@ def _run_op(a, *args, **kwargs):
 Parameter._OverloadAllOperators()
 tf.register_tensor_conversion_function(Parameter, lambda x, *args, **kwds: x.read_value())
 
+TensorLike = Union[np.ndarray, tf.Tensor, tf.Variable, Parameter]
+"""
+Type alias for tensor-like types that are supported by most TensorFlow, NumPy and GPflow operations.
+
+NOTE: Does not work with multipledispatch dispatchers. See `TensorLikeTypes`.
+"""
```
Comment (st--) [inline gpflow/base.py:312]:
```suggestion
# Type alias for tensor-like types that are supported by most TensorFlow, NumPy and GPflow operations.
#
# NOTE: The multipledispatch package uses tuples, not typing.Union, for dispatching on multiple types.
# For calling a Dispatcher's register() method (e.g. the `@conditional.register` decorator), use `TensorLikeTypes` instead
```

File: gpflow/base.py [hunk 3]
```diff
@@ -307,6 +304,16 @@ def _run_op(a, *args, **kwargs):
...
+
+TensorLikeTypes = (np.ndarray, tf.Tensor, tf.Variable, Parameter)
+""" Collection of tensor-like types for use with multipledispatch dispatchers.  """
```
Comment (st--) [inline gpflow/base.py:315]:
```suggestion
# Collection of tensor-like types for registering implementations with multipledispatch dispatchers.
```

File: gpflow/base.py (hunk 2 above)
Comment (joelberkeley-pio) [reply gpflow/base.py:312]: i chose docstrings so that it would appear in the docs. users may want to use these types in their own code

Comment (st--) [reply gpflow/base.py:312]: Will we have to add `autodata` or some such to the API documentation for this to actually get pulled in?
I just looked it up as I didn't know it handled module-constant docstrings like this; Sphinx also supports `#:` instead which I think might be more readable when just reading the source code?

File: gpflow/base.py [hunk 4]
```diff
@@ -307,6 +304,18 @@ def _run_op(a, *args, **kwargs):
...
+"""
+Type alias for tensor-like types that are supported by most TensorFlow, NumPy and GPflow operations.
+
+NOTE: Does not work with multipledispatch dispatchers. See `TensorLikeTypes`.
+"""
+
+# We've left this as object until we've tested the performance consequences of using the full set
+# see github issue #1434
+TensorLikeTypes = (object,)  # (np.ndarray, tf.Tensor, tf.Variable, Parameter)
+""" Collection of tensor-like types for use with multipledispatch dispatchers.  """
```
Comment (st--) [inline gpflow/base.py:317]:
```suggestion
#: Collection of tensor-like types for registering implementations with multipledispatch dispatchers.
TensorLikeTypes = (object,)  # (np.ndarray, tf.Tensor, tf.Variable, Parameter)
# We've left this as object until we've tested the performance consequences of using the full set
# see https://github.com/GPflow/GPflow/issues/1434
```

File: gpflow/base.py [hunk 5]
```diff
@@ -307,6 +304,18 @@ def _run_op(a, *args, **kwargs):
 Parameter._OverloadAllOperators()
 tf.register_tensor_conversion_function(Parameter, lambda x, *args, **kwds: x.read_value())
 
+TensorLike = Union[np.ndarray, tf.Tensor, tf.Variable, Parameter]
+"""
+Type alias for tensor-like types that are supported by most TensorFlow, NumPy and GPflow operations.
+
+NOTE: Does not work with multipledispatch dispatchers. See `TensorLikeTypes`.
+"""
```
Comment (st--) [inline gpflow/base.py:312]:
```suggestion
#: Type alias for tensor-like types that are supported by most TensorFlow, NumPy and GPflow operations.
#:
#: NOTE: The multipledispatch package uses tuples, not typing.Union, for dispatching on multiple types.
#: For calling a Dispatcher's register() method (e.g. the `@conditional.register` decorator), use `TensorLikeTypes` instead
TensorLike = Union[np.ndarray, tf.Tensor, tf.Variable, Parameter]
```

File: gpflow/base.py (hunk 2 above)
Comment (joelberkeley-pio) [reply gpflow/base.py:312]: meh - is that needed here?

File: gpflow/base.py (hunk 4 above)
Comment (joelberkeley-pio) [reply gpflow/base.py:318]: docstrings are usually below the item, no?

Comment (st--) [reply gpflow/base.py:318]: strings after, comments before (or same line): https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autodata

Comment (joelberkeley-pio) [reply gpflow/base.py:318]: ok. either way, is there a problem with using """ xyz """? I prefer that

Comment (joelberkeley-pio) [reply gpflow/base.py:318]: pycharm renders `"""` differently to `#:`, and I am (and I would wager others are) more used to `#` meaning comment not docstring, even if it has a `:` after it.

also multiline docstrings are common and easier with `"""`

File: gpflow/base.py [hunk 6]
```diff
@@ -307,6 +304,19 @@ def _run_op(a, *args, **kwargs):
...
+TensorLike = Union[np.ndarray, tf.Tensor, tf.Variable, Parameter]
+"""
+Type alias for tensor-like types that are supported by most TensorFlow, NumPy and GPflow operations.
+
+NOTE: Union types like this do not work with the `register` method of multipledispatch's
+`Dispatcher` class. Instead use `TensorLikeTypes` for dispatching on tensor-like types.
+"""
+
+# We've left this as object until we've tested the performance consequences of using the full set
+# see github issue #1434
```
Comment (st--) [inline gpflow/base.py:316]:
```suggestion
# see https://github.com/GPflow/GPflow/issues/1434
```
I think it's generally more helpful to have the full link there - one click instead of a bunch of clicking & typing

File: gpflow/base.py (hunk 4 above)
Comment (st--) [reply gpflow/base.py:318]: I prefer the comments but I don't mind very much :shrug:
I've left the content change as a new separate suggestion

**Comment by st--** [comment]
> This is the risk that one has to accept by using third party utilities over the stdlib

I wish Python would provide multiple dispatch by default (in the stdlib)...

Did you already check whether sphinx pulls in the module level docstrings even without `autodata`?

File: gpflow/base.py [hunk 7]
```diff
@@ -307,6 +304,19 @@ def _run_op(a, *args, **kwargs):
 Parameter._OverloadAllOperators()
 tf.register_tensor_conversion_function(Parameter, lambda x, *args, **kwds: x.read_value())
 
+TensorLike = Union[np.ndarray, tf.Tensor, tf.Variable, Parameter]
```
Comment (st--) [inline gpflow/base.py:307]:
```suggestion

TensorLike = Union[np.ndarray, tf.Tensor, tf.Variable, Parameter]
```
(to visually set it more apart from the Parameter class details; black is happy with two-line blanks in this context)

File: gpflow/base.py (hunk 6 above)
Comment (joelberkeley-pio) [reply gpflow/base.py:316]: ok. will do

**Comment by joelberkeley-pio** [comment]
> I just realised that this is breaking backwards compatibility, if anyone actually used `TensorLike` to mean `object` before (e.g. in registering their own implementations for multipledispatch)... how do you think we should handle this ?

We can either:

  * class it as a bug fix (which is reasonable), and just take the flak for changing it
  * use `TensorLike` to refer to the tuple of `(np.ndarray, ...)`, so the change wouldn't be less likely to impact users, and have a different name for the `Union`

Other ideas?

File: gpflow/base.py (hunk 7 above)
Comment (joelberkeley-pio) [reply gpflow/base.py:307]: will do

**Comment by joelberkeley-pio** [comment]
e.g.
```
TensorType = Union[np.ndarray, tf.Tensor, tf.Variable, Parameter]
TensorLike = (object,)  # (np.ndarray, tf.Tensor, tf.Variable, Parameter)
```
or instead of `TensorType`

File: gpflow/base.py (hunk 3 above)
Comment (joelberkeley-pio) [reply gpflow/base.py:318]: will do

File: gpflow/base.py [hunk 8]
```diff
@@ -308,6 +306,24 @@ def _run_op(a, *args, **kwargs):
 tf.register_tensor_conversion_function(Parameter, lambda x, *args, **kwds: x.read_value())
 
 
+TensorType = Union[np.ndarray, tf.Tensor, tf.Variable, Parameter]
+"""
+Type alias for tensor-like types that are supported by most TensorFlow, NumPy and GPflow operations.
+
+NOTE: Union types like this do not work with the `register` method of multipledispatch's
+`Dispatcher` class. Instead use `TensorLikeTypes` for dispatching on tensor-like types.
```
Comment (st--) [inline gpflow/base.py:314]:
```suggestion
`Dispatcher` class. Instead use `TensorLike` for dispatching on tensor-like types.
```

File: gpflow/base.py [hunk 9]
```diff
@@ -308,6 +306,24 @@ def _run_op(a, *args, **kwargs):
...
+"""
+
+
+# We've left this as object until we've tested the performance consequences of using the full set
+# (np.ndarray, tf.Tensor, tf.Variable, Parameter), see https://github.com/GPflow/GPflow/issues/1434
+TensorLike: Final[Tuple[type, ...]] = (object,)
+"""
+:var TensorLike: Collection of tensor-like types for registering implementations with
```
Comment (st--) [inline gpflow/base.py:322]:
```suggestion
:const TensorLike: Collection of tensor-like types for registering implementations with
```
?

File: setup.py [hunk 10]
```diff
@@ -18,7 +18,13 @@
 
 
 # Dependencies of GPflow
-requirements = ["numpy>=1.10.0", "scipy>=0.18.0", "multipledispatch>=0.6", "tabulate"]
+requirements = [
+    "numpy>=1.10.0",
+    "scipy>=0.18.0",
+    "multipledispatch>=0.6",
+    "tabulate",
+    "typing_extensions",
```
Comment (st--) [inline setup.py:26]:
```suggestion
    "typing_extensions",  # for python 3.6 compatibility
```

File: gpflow/base.py (hunk 9 above)
Comment (joelberkeley-pio) [reply gpflow/base.py:322]: nice. didn't know about that one

File: gpflow/base.py (hunk 8 above)
Comment (joelberkeley-pio) [reply gpflow/base.py:314]: good catch

File: gpflow/base.py (hunk 9 above)
Comment (joelberkeley-pio) [reply gpflow/base.py:322]: oh, I think it's only for _referencing_ not declaring ...

for const
>     Reference a “defined” constant. This may be a Python variable that is not intended to be changed.

for var
> var, ivar, cvar: Description of a variable.

though it's not particularly clear

Comment (joelberkeley-pio) [reply gpflow/base.py:322]: i noticed cos pycharm doesn't highlight `const` like it highlights `var`
